webhooks/unifiedstoragequota/validation/unifiedstoragequota_validator.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	return handleRequest(webhookRequestContext)
+}
+
+func (h *RequestedCapacityHandler) HandleCreate(ctx *pkgctx.WebhookRequestContext) CapacityResponse {
+	vm := &vmopv1.VirtualMachine{}
+	if err := h.Converter.FromUnstructured(ctx.Obj.UnstructuredContent(), vm); err != nil {
+		return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+	}
+
+	scName := vm.Spec.StorageClass
+
+	sc := &storagev1.StorageClass{}
+	if err := h.Client.Get(ctx, client.ObjectKey{Name: scName}, sc); err != nil {
+		if apierrors.IsNotFound(err) {
+			return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+		}
+		return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+	}
+
+	var capacity *resource.Quantity
+
+	if vm.Status.Storage == nil || !conditions.IsTrue(vm, vmopv1.VirtualMachineConditionCreated) {
+		if vm.Spec.Image == nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, errors.New("vm.Spec.Image is required"))}
+		}
+		vmiName := vm.Spec.Image.Name
+		var imageStatus vmopv1.VirtualMachineImageStatus
+
+		switch vm.Spec.Image.Kind {
+		case "VirtualMachineImage":
+			vmi := &vmopv1.VirtualMachineImage{}
+			if err := h.Client.Get(ctx, client.ObjectKey{Namespace: vm.Namespace, Name: vmiName}, vmi); err != nil {
+				if apierrors.IsNotFound(err) {
+					return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+				}
+				return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+			}
+			imageStatus = vmi.Status
+		case "ClusterVirtualMachineImage":
+			cvmi := &vmopv1.ClusterVirtualMachineImage{}
+			if err := h.Client.Get(ctx, client.ObjectKey{Name: vmiName}, cvmi); err != nil {
+				if apierrors.IsNotFound(err) {
+					return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+				}
+				return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+			}
+			imageStatus = cvmi.Status
+		}
+
+		if len(imageStatus.Disks) < 1 || imageStatus.Disks[0].Capacity == nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, errors.New("boot disk not found in image status"))}
+		}
+		capacity = imageStatus.Disks[0].Capacity
+	} else {
+		capacity = resource.NewQuantity(0, resource.BinarySI)
+
+		if vm.Status.Storage.Unshared != nil {
+			capacity.Add(*vm.Status.Storage.Unshared)
+		}
+		// Subtract the PVC usage/limit from the total used value as PVCs
+		// are counted separately.
+		for _, volume := range vm.Status.Volumes {
+			if volume.Type == vmopv1.VirtualMachineStorageDiskTypeManaged {
+				if volume.Used != nil {
+					capacity.Sub(*volume.Used)
+				}
+			}
+		}
+	}
+
+	return CapacityResponse{
+		RequestedCapacity: RequestedCapacity{
+			Capacity:         *capacity,
+			StorageClassName: scName,
+			StoragePolicyID:  sc.Parameters[scParamStoragePolicyID],
```

Comment:
Add a check on `sc.Parameters` to avoid potential NPE (since it's an [optional field](https://github.com/kubernetes/api/blob/master/storage/v1/types.go#L46-L48))?

Reply:
`sc.Parameters` is a map, so this wouldn't cause a panic for nil pointer dereference, right? this would just return an empty string. so the question is what is the desired behavior in this situation? what would be the reason this is not populated, and is it better to error in this situation or to return as is and let the Storage Policy Quota webhook fail if this is indeed an error condition?

Reply:
Ah, I got mixed up about reading and writing a nil map in Go, and only the latter would cause a NPE. I am fine with returning the empty value as is in that case. Thanks for adding a comment there!